analysis_scripts/analyze_quote_data/first_v_last.md

```diff
@@ -0,0 +1,261 @@
+    ## 2      doi:10.1007/128_2010_70 naturenews_citations
+    ## 3      doi:10.1007/164_2016_82 naturenews_citations
+    ## 4  doi:10.1007/3-540-26573-2_9 naturenews_citations
+    ## 5 doi:10.1007/3-540-28221-1_12 naturenews_citations
+    ## 6 doi:10.1007/3-540-44948-5_10 naturenews_citations
+
+## match quotes to authors
+
+``` r
+# first we only care about files that both have 
+# at least one quote and springer indexed citation
+file_id_intersect = intersect(full_quote_df$file_id, 
+                                name_info_df$file_id)
+
+# go through the files that meet the above criteria
+# and match citations to quoted speakers
+quote_author_df = NA
+missed_names_df = NA
+for(curr_id in file_id_intersect){
+    
+    curr_quote = subset(full_quote_df, file_id == curr_id)
+    curr_author = subset(name_info_df, file_id == curr_id)
+    
+    # we must make approx. name matches
+    # so we go through each quoted speaker, 
+    # looking for the best cited author match, IF it exists
+    names_df = subset(curr_quote, select=c("est_speaker"))
+    names_df$author = NA
+    for(idx in 1:nrow(names_df)){
+        curr_text = names_df$est_speaker[idx]
+        
+        # remove punctuations just in case
+        curr_text = gsub('[[:punct:] ]+', '', curr_text)
+        
+        # get_matched_string tries to find the name with the 
+        # closest substring with no penalty for deletions.
+        # Maximum number of mismatches is 3
+        names_df$author[idx] = get_matched_string(curr_text, 
+                                                  curr_author$author, 
+                                                  max_cost = 3)
+
+    }
+    # only keep the names we were able to find
+    curr_missed_names = subset(names_df, is.na(author))
+    names_df = na.omit(names_df)
+    
+    curr_quote = merge(curr_quote, names_df)
+    merged_df = merge(curr_quote[,c("author", "year", "est_speaker", "type", "file_id", "quote")], 
+                      curr_author[,c("author", "doi", "author_pos")],
+                      by = "author")
+    quote_author_df = rbind(quote_author_df, merged_df)
+    
+    if(nrow(curr_missed_names) > 0){
+        missed_names_df = rbind(missed_names_df, curr_missed_names)
+    }
+
+    
+}
+quote_author_df = quote_author_df[-1,]
+quote_author_df = unique(quote_author_df)
+
+missed_names_df = missed_names_df[-1,]
+```
+
+## QC: look into all missed quotes
+
+Missing names (i.e. names that are quoted, but not found in the citations) can occur for the following reasons:
+
+1.  The parsed name from the quote is incorrect
+
+2.  The quote is commentary, i.e. there is no citation needed for this quote
+
+3.  The quote is not from the first or last author
+
+4.  The quote is from a cited author, but the cited article is not indexed by springer
+
+Unfortunately, there isn't an easy way for me to know exactly which of the reasons a name is missing. We just hope there is no confounders.
```

Comment:
I see that you are using `max_cost = 3`. Why did you choose that value? Is there a way to see, for a particular quote, what's the best match no matter the cost (the top one)? I imagine it could be a way to debug the reasons a name is missing (what you mention below).

Reply:
Another way to say this would be: do you have examples of missed names? or pairs of quotes and authors that you expect your code to find? are you sure the code is getting those? If these make sense, I would create some unit tests for some of these functions with examples of expected outputs.

Reply:
yes, this is a good point. I will have to make a benchmark dataset for this. I was already starting to make a "final" benchmark dataset for all of the analyses that I plan to show in the end, so I think this is a good suggestion to really get me to finish this new benchmark.
I won't address in this PR, but I added it as an issue here: https://github.com/greenelab/nature_news_disparities/issues/24